openscm/scmdataframe/base.py

```diff
@@ -0,0 +1,1417 @@
+                                    ...
+        2008-05-01 00:00:00       9.329380
+        2008-06-01 00:00:00       9.414234
+        2008-07-01 00:00:00       9.496350
+        2008-08-01 00:00:00       9.581204
+        2008-09-01 00:00:00       9.666058
+        2008-10-01 00:00:00       9.748175
+        2008-11-01 00:00:00       9.833029
+        2008-12-01 00:00:00       9.915146
+        2009-01-01 00:00:00      10.000000
+        [109 rows x 1 columns]
+
+
+        Note that the values do not fall exactly on integer values as not all years
+        are exactly the same length.
+
+        References
+        ----------
+        See the pandas documentation for
+        `resample <http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.resample.html>`
+        for more information about possible arguments.
+        """
+        orig_dts = self["time"]
+        target_dts = generate_range(
+            orig_dts.iloc[0], orig_dts.iloc[-1], to_offset(rule)
+        )
+        return self.interpolate(list(target_dts), **kwargs)
+
+    def process_over(
+        self, cols: Union[str, List[str]], operation: str, **kwargs: Any
+    ) -> pd.DataFrame:
+        """
+        Process the data over the input columns.
+
+        Parameters
+        ----------
+        cols
+            Columns to perform the operation on. The timeseries will be grouped
+            by all other columns in ``self.meta``.
+
+        operation : ['median', 'mean', 'quantile']
+            The operation to perform. This uses the equivalent pandas function.
+            Note that quantile means the value of the data at a given point in the
+            cumulative distribution of values at each point in the timeseries, for
+            each timeseries once the groupby is applied. As a result, using
+            ``q=0.5`` is is the same as taking the median and not the same as
+            taking the mean/average.
+
+        **kwargs
+            Keyword arguments to pass to the pandas operation.
+
+        Returns
+        -------
+        :obj:`pd.DataFrame`
+            The quantiles of the timeseries, grouped by all columns in ``self.meta``
+            other than ``cols``
+
+        Raises
+        ------
+        ValueError
+            If the operation is not one of ['median', 'mean', 'quantile'].
+        """
+        cols = [cols] if isinstance(cols, str) else cols
+        ts = self.timeseries()
+        group_cols = list(set(ts.index.names) - set(cols))
+        grouper = ts.groupby(group_cols)
+
+        if operation == "median":
+            return grouper.median(**kwargs)
+        if operation == "mean":
+            return grouper.mean(**kwargs)
+        if operation == "quantile":
+            return grouper.quantile(**kwargs)
+
+        raise ValueError("operation must be one of ['median', 'mean', 'quantile']")
+
+    def convert_unit(self, unit: str, inplace=False, **kwargs: Any) -> ScmDataFrameBase:
```

Comment:
can we add a keyword, `context: Optional[str] = None`, which gets passed through to the `UnitConverter`?